### PR TITLE
Render user comments as docs of nodes in pipeline

### LINKED ISSUE #1358: Pass through pipeline comments that are associated with nodes
**Is your feature request related to a problem? Please describe.**

As a user, I can attach comments to nodes that describe its purpose.

image

Elyra currently does not pass through this information to the target environment (Kubeflow Pipelines and Apache Airflow), even though it might be useful to the user. Apache Airflow, for example, explicitly supports documentation for DAGs and tasks: https://airflow.apache.org/docs/apache-airflow/stable/tutorial.html#adding-dag-and-tasks-documentation

**Describe the solution you'd like**
If a user associated a comment with a node, the comment text is passed through to pipeline processors that support node-level documentation.

### PR DESCRIPTION
**Pull request by yuhuishi-convect**
Closes #1358 

### What changes were proposed in this pull request?

* This feature associates user comment blocks to nodes in pipelines, as requested in #1358. 
* The comment blocks will be rendered as 
  * Pod annotations in KFP pipelines
  * `doc` field of a task in Airflow pipelines 
* Implementation details
  * An optional `doc` field is added to `Operation` class
  * During parsing phase, for each node in a pipeline_def we search for if there is any user comment block associated with it. If so, we set `doc` of the created `operation` to user comments (in case there are multiple comment blocks associated with one node, we concat them)
  * During processing phase, we add a pod annotation or add `op.doc="xxx"` in KFP and Airflow processors correspondingly, if an operation has a `doc` field.

### How was this pull request tested?

* `make test-server`
* Added one pipeline JSON with user comment blocks `elyra/tests/pipeline/resources/sample_pipelines/pipeline_3_node_sample_with_comments.json`; a test (`elyra/tests/pipeline/test_pipeline_parser.py:test_pipeline_with_comments`) that verifies if `doc` field is correctly populated by the parser.
* Started a jupyterlab and tested exporting a pipeline with user comments to Airflow and KFP pipeline files.

<a href="https://www.loom.com/share/0b3100d0e7034537b9d0f20a66f510e6">
    <p>pipeline-wit… - JupyterLab - 18 January 2022 - Watch Video</p>
    
  </a>
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: elyra/pipeline/pipeline_definition.py [hunk 1]
```diff
@@ -460,6 +468,27 @@ def get_node(self, node_id: str):
...
+        Given a node id returns the assoicated comments in the pipeline
+        :param node_id: the node id
+        :return: the comments or None
+        """
+        comments = []
+
+        for pipeline in self.pipelines:
+            comment_list = pipeline.comments
+            for comment in comment_list:
+                assoicated_node_id_list = comment.get("associated_id_refs", [])
```
Comment (kevin-bates) [inline elyra/pipeline/pipeline_definition.py:482]: Typo...
```suggestion
                associated_node_id_list = comment.get("associated_id_refs", [])
```
(and its reference below)

**Review by kevin-bates — APPROVED** [review]
Hi @yuhuishi-convect - thank you for the great contribution.  I'll defer to others for actual integration details, but, from the code perspective and addressing the issue, this looks good to me.

Thank you for the great PR content as well.

File: elyra/pipeline/pipeline_definition.py [hunk 2]
```diff
@@ -460,6 +468,27 @@ def get_node(self, node_id: str):
...
+        :param node_id: the node id
+        :return: the comments or None
+        """
+        comments = []
+
+        for pipeline in self.pipelines:
+            comment_list = pipeline.comments
+            for comment in comment_list:
+                associated_node_id_list = comment.get("associated_id_refs", [])
+                for ref in assoicated_node_id_list:
```
Comment (kevin-bates) [inline elyra/pipeline/pipeline_definition.py:483]: (Sorry, there was a reference to the fixed typo so just committing the suggestion wasn't sufficient (until now :smile:))
```suggestion
                for ref in associated_node_id_list:
```

Comment (yuhuishi-convect) [reply elyra/pipeline/pipeline_definition.py:483]: thanks @kevin-bates , excellent catch

File: elyra/pipeline/kfp/processor_kfp.py [hunk 3]
```diff
@@ -532,6 +532,9 @@ def _cc_pipeline(self,
                                                                  .format(pipeline_envs['ELYRA_WRITABLE_CONTAINER_DIR'])  # noqa
                                                          })
 
+                if operation.doc:
+                    target_ops[operation.id].add_pod_annotation('elyra/user-doc', operation.doc)
```
Comment (ptitzler) [inline elyra/pipeline/kfp/processor_kfp.py:536]: Could we use `elyra/node-user-doc` to make it clear that the doc string is associated with the node/pod and not the pipeline? Other annotations, such as `elyra/node-name` and `elyra/node-type` already follow that naming convention.

```suggestion
                    target_ops[operation.id].add_pod_annotation('elyra/node-user-doc', operation.doc)
```

**Review by ptitzler — CHANGES_REQUESTED** [review]
Nice work! Overall this works fine with one caveat: the behavior is inconsistent for empty comments: these empty strings (or strings containing only whitespaces) should be ignored because they provide no value to the user.
Current behavior for these scenarios:
 - user associates an empty comment with a node -> no "doc string" is generated (as expected and desired!)
 - user associates two or more empty comments with a node -> "doc string" is generated even though there is nothing to document
 - user associates two or more comments with a node (one of the comments being empty) -> "doc string" is generated, including an "extra" newline for the empty comment.

**Review by ptitzler — APPROVED** [review]
LGTM

**Review by akchinSTC — APPROVED** [review]
tested locally and LGTM. 
Looks like theres some kind of intermittent test failure going on, re-ran the CI tests and seems to have resolved itself.

**Pull request merged**